src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,199 @@
+    if (subscribers.size() == 0) {
+      objects_to_subscribers_.erase(subscribers_it);
+    }
+  }
+  subscribers_to_objects_.erase(subscribing_objects_it);
+  return 1;
+}
+
+int SubscriptionIndex::EraseEntry(const ObjectID &object_id,
+                                  const NodeID &subscriber_id) {
+  // Erase from subscribers_to_objects_;
+  auto subscribers_to_objects_it = subscribers_to_objects_.find(subscriber_id);
+  if (subscribers_to_objects_it == subscribers_to_objects_.end()) {
+    return 0;
+  }
+  auto &objects = subscribers_to_objects_it->second;
+  auto object_it = objects.find(object_id);
+  if (object_it == objects.end()) {
+    RAY_CHECK(objects_to_subscribers_.count(object_id) == 0);
+    return 0;
+  }
+  objects.erase(object_it);
+  if (objects.size() == 0) {
+    subscribers_to_objects_.erase(subscribers_to_objects_it);
+  }
+
+  // Erase from objects_to_subscribers_.
+  auto objects_to_subscribers_it = objects_to_subscribers_.find(object_id);
+  // If code reaches this line, that means the object id was in the index.
+  RAY_CHECK(objects_to_subscribers_it != objects_to_subscribers_.end());
+  auto &subscribers = objects_to_subscribers_it->second;
+  auto subscriber_it = subscribers.find(subscriber_id);
+  // If code reaches this line, that means the subscriber id was in the index.
+  RAY_CHECK(subscriber_it != subscribers.end());
+  subscribers.erase(subscriber_it);
+  if (subscribers.size() == 0) {
+    objects_to_subscribers_.erase(objects_to_subscribers_it);
+  }
+  return 1;
+}
+
+bool Subscriber::Connect(LongPollConnectCallback long_polling_reply_callback) {
+  if (long_polling_reply_callback_ == nullptr) {
+    long_polling_reply_callback_ = long_polling_reply_callback;
+    return true;
+  }
+  return false;
+}
+
+void Subscriber::QueueMessage(const ObjectID &object_id, bool try_publish) {
+  mailbox_.push_back(object_id);
+  if (try_publish) {
+    PublishIfPossible();
+  }
+}
+
+bool Subscriber::PublishIfPossible(bool force) {
+  if (long_polling_reply_callback_ == nullptr) {
+    return false;
+  }
+  if (force || mailbox_.size() > 0) {
+    long_polling_reply_callback_(mailbox_);
+    long_polling_reply_callback_ = nullptr;
+    mailbox_.clear();
+    return true;
+  }
+  return false;
+}
+
+void PubsubCoordinator::Connect(const NodeID &subscriber_node_id,
+                                LongPollConnectCallback long_poll_connect_callback) {
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Long polling connection is initiated by " << subscriber_node_id;
+  RAY_CHECK(long_poll_connect_callback != nullptr);
+
+  if (is_node_dead_(subscriber_node_id)) {
+    std::vector<ObjectID> empty_ids;
+    // Reply to the request so that it is properly GC'ed.
+    long_poll_connect_callback(empty_ids);
+    return;
+  }
+
+  if (subscribers_.count(subscriber_node_id) == 0) {
+    subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>());
+  }
+  auto subscriber = subscribers_[subscriber_node_id];
```

Comment:
Nit: Using this `.count()`/`.emplace()`/lookup pattern is supposedly a bit of an antipattern, since it involves multiple lookups.
```suggestion
  auto it = subscribers_.find(subscriber_node_id);
  if (it == subscribers_.end()) {
    it = subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>()).first;
  }
  auto subscriber = it->second;
```

Reply:
Ah good point. I just hate to use this `it` based code... but it seems like I should just use that.

Reply:
Embrace the iterators, it's the C++ way! 😁